Tail.scala

```diff
@@ -0,0 +1,226 @@
+                                    -> MatVecMul -> OUTPUT
+
+Notes from Yue:
+  Probably don't need deserialize at all
+    just need to create a buffer of size n=# of points for FFT
+      Deserialize does this but with MIMO prefix/suffix that we don't need here
+
+  Don't need anything after unscramble. Delete it!
+
+  This uses chisel3.dsptools and that will be deprecated soon
+    don't worry about this for now!
+
+
+Notes from 10/29:
+  1. Convert Tail to a lazy module
+    - create a new module LazyTail
+        - `extends LazyModule` instead of `extends Module` to make it a lazy module (with a diplomatic section)
+    - Regmap should live inside this new lazy module
+        - should wire it to the input and output of the Tail module
+
+    - put everything existing in the LazyModuleImp (non-diplomatic section)
+    - diplomatic section is going to be the TLRegisterNode (this specifies the base memory address and connects it to main mem -- example: https://github.com/ucberkeley-ee290c/fa18-gps/blob/50189c74f92dddaf88ba9a8ac322d4e21f140746/src/main/scala/gps/FFT/WriteQueueBlock.scala#L110)
+*/
+
+package fftgenerator
+
+import chisel3._
+import chisel3.experimental.FixedPoint
+import chisel3.util._
+import chisel3.util.{Decoupled, Counter}
+import dsptools.numbers._
+import freechips.rocketchip.diplomacy.LazyModule
+import freechips.rocketchip.subsystem.BaseSubsystem
+
+import freechips.rocketchip.regmapper._
+import freechips.rocketchip.tilelink.TLRegisterNode
+import freechips.rocketchip.config.Parameters
+import freechips.rocketchip.diplomacy._
+
+import breeze.math.Complex
+import chisel3.experimental._
+import chisel3.ExplicitCompileOptions
+import chisel3.internal.firrtl.KnownBinaryPoint
+import craft._
+import dsptools._
+import dsptools.numbers.implicits._
+import dspjunctions._
+import dspblocks._
+import scala.math._
+import scala.math.sqrt
+import scala.collection.mutable.ListBuffer
+
+trait TailParams[T <: Data] extends DeserializeParams[T] with FFTConfig[T] with UnscrambleParams[T] {
+    // Number of lanes in FFT. Should be same as N.
+    val lanes: Int
+    // n-point FFT
+    val n: Int
+    val S: Int
+}
+
+/* This is already defined in the Spine
+trait FileNameParams{
+  val FileNames: Seq[String]
+}*/
+
+// todo: FFT uses DspComplex which is from dsptools and that's getting deprecated. Can we move DspComplex/fixedPoint/etc
+// directly into FFT from there?
+// todo: make a list of everything that uses dsptools. See if that can be moved directly into FFT repo
+// ANSWER: nontrivial amount of dependencies so can't just copy-paste
+    // uses complex numbers, fixedpoint, dsptools computation tools for those (adding stages to multiplication etc)
+
+case class FixedTailParams(
+    IOWidth: Int = 16,
+    binaryPoint: Int = 8,
+    lanes: Int = 2, // todo ask yue: why are N and lanes different? -- can remove N, only lanes is used by provided code
+    n: Int = 2,
+    S: Int = 256,
+    pipelineDepth: Int = 0,
```

Comment:
In the documentation, all of these parameters should be explained. If not, then you need to write some documentation to explain this.

Reply:
Where should I put the documentation?